### PR TITLE
Improve WiFi Resilience and Flexibility

### PR DESCRIPTION
**Pull request by sellensr**
Scope: Add an optional fail_fast parameter to run() and mqttStatus() to allow them to return promptly if desired by the calling sketch.

Detect a wifi failure and attempt to reconnect from inside AIO unless fail_fast is true.

In the WiFi code, test that SSID is a non-empty string before trying to connect. Disconnect any existing wifi connection and pause long enough for it to take effect before going ahead.

References:
https://github.com/adafruit/Adafruit_IO_Arduino/issues/99
https://github.com/adafruit/Adafruit_IO_Arduino/issues/102

Limitations: There should be no change in the operation of existing code, except that the initial WiFi connection will take 300 ms longer to wait for the disconnect to take effect. The included example is specific to the Feather WINC1500 platform because it is an example for implementing platform specific code to manage the WiFi outside AIO. I have not tested the modification on all the supported hardware.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,3 +12,8 @@ platformio.ini
 
 # visual studio code
 .vscode
+
+.development
```
Comment (brentru) [inline .gitignore]: Please remove

File: examples/adafruitio_25_self_set_wifi/config.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/************************ Adafruit IO Config *******************************/
```
Comment (brentru) [inline examples/adafruitio_25_self_set_wifi/config.h]: Ok, please remove this example. I'll test it on M0/ESP32/ESP8266 Targets.

File: examples/adafruitio_25_self_set_wifi/adafruitio_25_self_set_wifi.ino [hunk 3]
```diff
@@ -0,0 +1,89 @@
+// Adafruit IO Publish Example with Independent WINC1500 WiFi
```
Comment (brentru) [inline examples/adafruitio_25_self_set_wifi/adafruitio_25_self_set_wifi.ino]: Ok, please remove this example. I'll test it on M0/ESP32/ESP8266 Targets.

File: src/wifi/AdafruitIO_AIRLIFT.h [hunk 4]
```diff
@@ -158,6 +158,10 @@ class AdafruitIO_AIRLIFT : public AdafruitIO {
     /**************************************************************************/
     void _connect()
     {
+      if(strlen(_ssid) != 0)
```
Comment (brentru) [inline src/wifi/AdafruitIO_AIRLIFT.h]: Could you please move your SSID work into a separate branch and open a separate PR for it?

It helps narrow the scope if any issues crop up during tests or releases.

Comment (sellensr) [reply src/wifi/AdafruitIO_AIRLIFT.h]: Changes made and SSID commit reverted

**Comment by sellensr** [comment]
I think my descriptions are correct, but there are parts of the library I haven't used, e.g. times, dashboards, and groups.

**Comment by brentru** [comment]
Wow, thank you for the doxygen descriptions you've made. This looks ready to merge and release.
